src/persist-client/src/impl/encoding.rs

```diff
@@ -499,6 +499,12 @@ pub struct SerdeLeasedBatch(Vec<u8>);
 
 impl<T: Timestamp + Codec64> From<LeasedBatch<T>> for SerdeLeasedBatch {
     fn from(mut x: LeasedBatch<T>) -> Self {
+        // Make sure to encode as proto first, to capture the current state of
+        // "droppable". Otherwise, all batches that went through
+        // SerdeLeasedBatch once get forever marked as droppable, which is
+        // incorrect.
+        let result = SerdeLeasedBatch(x.into_proto().encode_to_vec());
```

Comment:
Ah, this makes more sense, though it is impossible to take a `SerdeLeasedBatch` into a droppable `LeasedBatch` https://github.com/MaterializeInc/materialize/pull/14177/commits/7e31d7224cb1d7f2cdd4707fe9d0afd3c88bec7f#diff-892645a5afe4772f7e97b71536e2b4ddfa8802b9c8f71d57e9bea34c7085d574R470

Reply:
You don't follow my comment on the code? OR Sean's comment about that?

Reply:
@danhhz I think there was a misunderstanding between Aljoscha and I here.
- The implementation of `From<SerdeLeasedBatch>` always generates an undroppable `LeasedBatch`.
- Aljoscha re-organized some code to ensure that we never serialized a droppable `LeasedBatch`.

I think Aljoscha wrote this bit of code without seeing that the `From` implementation is never droppable, which I was pointing out, while acknowledging that this code movement makes the intended behavior more obvious.